umpost/um2netcdf.py

```diff
@@ -734,6 +725,41 @@ def fix_level_coord(cube, z_rho, z_theta, tol=1e-6):
                 c_sigma.var_name = 'sigma_theta'
 
 
+def fix_pressure_levels(cube, decimals=5):
+    """
+    Reformat pressure level data for NetCDF output.
+
+    This converts units, rounds small fractional errors & ensures pressure is
+    decreasing (following the CMIP6 standard).
+
+    Parameters
+    ----------
+    cube : iris Cube (modifies in place)
+    decimals : number of decimals to round to
+
+    Returns
+    -------
+    None if cube lacks pressure coord or is modified in place, otherwise a new
+    cube if the pressure levels are reversed.
+    """
+    try:
+        pressure = cube.coord('pressure')
+    except iris.exceptions.CoordinateNotFoundError:
+        return
+
+    # update existing cube metadata in place
+    pressure.attributes['positive'] = 'down'
+    pressure.convert_units('Pa')
+
+    # Round small fractions otherwise coordinates are off by 1e-10 in ncdump output
+    pressure.points = np.round(pressure.points, decimals)
+
+    if pressure.points[0] < pressure.points[-1]:
+        # Flip to get pressure decreasing as per CMIP6 standard
+        # NOTE: returns a new cube!
+        return iris.util.reverse(cube, 'pressure')
```

Comment:
I noticed line 760 is using the entire `cube` in reversing the pressure levels:
https://github.com/ACCESS-NRI/um2nc-standalone/blob/dd9c18dd1ad709bd7c3dd839d3eb74bc6ebbfae4/umpost/um2netcdf.py#L757-L760

However, `iris.util.reverse()` takes `cube`s or `array`s:
https://github.com/SciTools/iris/blob/e1f5a8b880e443c58cba7770a1ffd32e067a64f2/lib/iris/util.py#L518-L524 (preview is failing....)

I'm experimenting with modifying the pressure array order in place, in the event the coord can be updated within the `cube`. It's currently breaking...

@MartinDix from your knowledge of `iris` do you know if modifying the `pressure.points` array is recommended practice?

```python
    pressure = cube.coord('pressure')

    # snip processing...

    if pressure.points[0] < pressure.points[-1]:
        # Flip to get pressure decreasing as per CMIP6 standard
        # return iris.util.reverse(cube, 'pressure')  # ORIG
        pressure.points = np.flip(pressure.points)  # TODO: is this valid???
```

Reply:
Just to jump in with some extra info, let me know if I've misunderstood the question though! My understanding is that the pressure coordinate's points ordering shouldn't change separately from of the rest of the cube's data.

My (rough) understanding of a typical `iris.cube` is that it holds an n-dimensional array as it's `data` attribute, and for each dimension has a separate coordinate array. The points in the coordinate arrays specify the locations corresponding to the main `data`. E.g. a cube describing surface temperature might have a 180x360 array of values, plus a length 180 `latitude` coordinate array, and length 360 longitude coordinate array. 

The temperature given by `cube.data[10,55]` would correspond to the location specified by `cube.coord("latitude")[10]` and `cube.coord("latitude")[55]`. But this means if we flip one of the coordinates independently of the data, the wrong locations will be given for the cube data.

It looks like `iris.util.reverse(cube, 'pressure')` is flipping both the `pressure` coordinate, and the actual cube's `data`, so that the two still line up:

```python
> cube_on_plev = cubes[142] # Get a cube with a pressure coordinate

# extract single latitude and longitude point, so that vertical coordinate pressure is the only coordinate
> single_cell_cube = cube_on_plev.extract(iris.Constraint(latitude=0.625, longitude = 0.9375))

# Print out initial pressure points and data array
> single_cell_cube.coord("pressure").points

array([   1.,    5.,   10.,   20.,   30.,   50.,   70.,  100.,  150.,
        200.,  250.,  300.,  400.,  500.,  600.,  700.,  850.,  925.,
       1000.])

> single_cell_cube.data
array([2.23973939e-04, 1.04892252e-02, 5.33754104e-02, 2.66307022e-01,
       9.27435023e-01, 6.87145581e+00, 3.06664567e+01, 5.38439087e+01,
       5.30945653e+01, 4.83872854e+01, 2.50941701e+01, 1.22997686e+01,
       9.74147291e+00, 2.95951569e+01, 5.42121837e+01, 5.50435870e+01,
       5.61263122e+01, 8.55014709e+01, 7.11272948e+01])

# Reverse cube according to pressure coordinate
> single_cell_cube_reverse = iris.util.reverse(single_cell_cube, single_cell_cube.coord("pressure"))

# Print out new data and pressure coordinate arrays
> single_cell_cube_reverse.coord("pressure").points
array([1000.,  925.,  850.,  700.,  600.,  500.,  400.,  300.,  250.,
        200.,  150.,  100.,   70.,   50.,   30.,   20.,   10.,    5.,
          1.])

>single_cell_cube_reverse.data
array([7.11272948e+01, 8.55014709e+01, 5.61263122e+01, 5.50435870e+01,
       5.42121837e+01, 2.95951569e+01, 9.74147291e+00, 1.22997686e+01,
       2.50941701e+01, 4.83872854e+01, 5.30945653e+01, 5.38439087e+01,
       3.06664567e+01, 6.87145581e+00, 9.27435023e-01, 2.66307022e-01,
       5.33754104e-02, 1.04892252e-02, 2.23973939e-04])

# Both have flipped!
```

Reply:
Ok, that makes sense.

I've skipped the test on account of problems at the um2nc & iris "boundary". Implementing a mocked `reverse()` requires faffing with return data structures OR using the real function & making a complicated custom cube arg. Both options are ugly for testing a _one liner_ call to an external library.

I can leave the skipped test in for future work OR delete it & comment why the pressure reversal isn't tested. Does anyone (@blimlim @marc-white  have a preference either way?

Reply:
I think the skipped test is fine for now, but add an issue to deal with it later. It definitely needs to be tested, but data transformations (even ones as basic as flipping) are a massive faff that need to be stepped through very methodically.

Reply:
#108 now covers the testing task.

I think deferring is a good step for now, to free effort for modularising & retrofitting tests to other functions.